packages/devtools_app/lib/src/charts/flutter/flame_chart.dart

```diff
@@ -14,6 +18,162 @@ const double sectionSpacing = 15.0;
 const double topOffset = rowHeightWithPadding;
 const double sideInset = 70.0;
 
+abstract class FlameChart<T, V extends FlameChartNodeDataMixin>
+    extends StatefulWidget {
+  const FlameChart(
+    this.data, {
+    @required this.duration,
+    @required this.height,
+    @required this.totalStartingWidth,
+    @required this.startInset,
+    @required this.selectionNotifier,
+    @required this.onSelection,
+  });
+
+  final T data;
+
+  final Duration duration;
+
+  final double totalStartingWidth;
+
+  final double height;
+
+  final double startInset;
+
+  final ValueListenable<V> selectionNotifier;
+
+  final void Function(V event) onSelection;
+
+  double get startingContentWidth =>
+      totalStartingWidth - startInset - sideInset;
+}
+
+abstract class FlameChartState<T extends FlameChart> extends State<T>
+    with AutoDisposeMixin {
+  LinkedScrollControllerGroup _linkedScrollControllerGroup;
+
+  List<FlameChartRow> rows;
+
+  @override
+  void initState() {
+    super.initState();
+    addAutoDisposeListener(widget.selectionNotifier);
+    _linkedScrollControllerGroup = LinkedScrollControllerGroup();
+  }
+
+  @override
+  void didUpdateWidget(T oldWidget) {
+    if (widget.data != oldWidget.data) {
+      _linkedScrollControllerGroup.resetScroll();
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    // Initialize the flame chart elements before building.
+    initFlameChartElements();
+
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return ListView.builder(
+          itemCount: rows.length,
+          itemBuilder: (context, index) {
+            return ScrollingFlameChartRow(
+              linkedScrollControllerGroup: _linkedScrollControllerGroup,
+              row: rows[index],
+              width: math.max(constraints.maxWidth, widget.totalStartingWidth),
+            );
+          },
+        );
+      },
+    );
+  }
+
+  void initFlameChartElements();
```

Comment:
My understanding is that this method will only serve to populate `rows`.

Consider making this a callback field on the widget, and making it return the list of rows.  That way, rows can be private.